src/challenge/ChallengeContext.ts

```diff
@@ -1,705 +1,61 @@
-            testResults: [{ outcome: true }],
-          });
-          this.actions["report-result"](
-            [{ outcome: true }],
-            code,
-            this.challenge.props.bookNode
-          );
-          this.actions["save-code"]({ code });
-        } else if (code && tests) {
-          this.actions["testpy"](code, tests);
-        }
-      }
-    },
-    "report-result": (
-      results: TestResults,
-      code: string,
-      bookNode: BookNodeModel
-    ) => {
-      let newTestOutcome =
-        results.length === 0
-          ? undefined
-          : results.filter((x) => x.outcome !== true).length === 0;
-      if (this.challenge.props.progressStorage) {
-        this.challenge.props.progressStorage.setResult(
-          bookNode,
-          newTestOutcome,
-          code
-        );
-      }
-    },
-    testpy: (code: string, tests: TestCases) => {
-      if (!this.challenge.worker) {
-        return;
-      }
-      if (this.challenge.state.editorState === ChallengeStatus.READY) {
-        if (this.challenge.interruptBuffer) {
-          this.challenge.interruptBuffer[0] = 0; // if interrupts are supported, just clear the flag for this execution
-        }
-
-        let additionalCode = "";
-
-        this.challenge.props.bookNode.additionalFiles?.forEach((file) => {
-          const escContents = this.challenge.state.additionalFilesLoaded[
-            file.filename
-          ]
-            .replace(/"/g, '\\"')
-            .replace(/\\/g, "\\\\");
-          additionalCode += `with open("${file.filename}", "w") as f:f.write("""${escContents}""")\n`;
-        });
-
-        const testsClone = structuredClone(tests);
-        let hasTurtleTest = false;
-
-        (testsClone || []).forEach((test) => {
-          if (test.out instanceof Array) {
-            test.out.forEach((out) => {
-              if (out.filename) {
-                out.filename =
-                  this.challenge.state.additionalFilesLoaded[out.filename];
-              }
-              if (out.typ === "t") {
-                hasTurtleTest = true;
-              }
-            });
-          }
-        });
-
-        if (hasTurtleTest) {
-          // switch to virtual mode
-          this.actions.awaitCanvas().then(() => {
-            this.challenge.canvasDisplayRef.current?.turtleReset(true);
-            this.challenge.worker?.postMessage({
-              cmd: "test",
-              code: code,
-              initCode: additionalCode,
-              tests: testsClone,
-              bookNode: this.challenge.props.bookNode,
-            });
-          });
-        } else {
-          // use the original tests for non turtle tests to avoid changing filenames to contents
-          this.challenge.worker.postMessage({
-            cmd: "test",
-            code: code,
-            initCode: additionalCode,
-            tests: tests,
-            bookNode: this.challenge.props.bookNode,
-          });
-        }
+type Actions = {
```

Comment:
simplified context class to a single interface. Previous functionality distributed between CodeRunner and Challenge